File: apps/desktop/src/main/kotlin/logger/AppErrorLogger.kt
// AppErrorLogger.kt — централизованное логирование ошибок Desktop.
//
// Записывает структурированные JSON-строки в ~/.messenger/logs/errors.log
// (macOS/Linux) или %APPDATA%\Messenger\logs\errors.log (Windows).
// При превышении 5 МБ ротирует файл (→ errors.log.old).
// Перехватывает непойманные исключения через UncaughtExceptionHandler.
// При старте приложения отправляет накопленные логи на /api/client-errors.
package logger

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put
import java.io.File
import java.net.HttpURLConnection
import java.net.URL
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.logging.Logger

object AppErrorLogger {

    private const val MAX_SIZE = 5 * 1024 * 1024L // 5 МБ
    private const val FLUSH_LINES = 100            // строк за один запрос

    private val scope = CoroutineScope(Dispatchers.IO)
    private val fmt   = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", Locale.US)
    private val jul   = Logger.getLogger("AppErrorLogger")
    private val json  = Json { ignoreUnknownKeys = true }

    private var logFile: File? = null

    // ─── Инициализация ─────────────────────────────────────────────────────────

    fun init() {
        val logsDir = resolveLogsDir().also { it.mkdirs() }
        logFile = File(logsDir, "errors.log")
        installCrashHandler()
    }

    // ─── Публичный API ──────────────────────────────────────────────────────────

    fun error(tag: String, message: String, throwable: Throwable? = null) {
        jul.severe("[$tag] $message${throwable?.let { "\n${it.stackTraceToString()}" } ?: ""}")
        append("error", tag, message, throwable?.stackTraceToString())
    }

    fun warn(tag: String, message: String) {
        jul.warning("[$tag] $message")
        append("warn", tag, message, null)
    }

    fun info(tag: String, message: String) {
        jul.info("[$tag] $message")
        append("info", tag, message, null)
    }

    /** Отправить накопленные логи на сервер (вызвать при старте приложения). */
    fun flushToServer(serverUrl: String) {
        if (serverUrl.isBlank()) return
        val file = logFile ?: return
        if (!file.exists() || file.length() == 0L) return
        scope.launch { doFlush(file, serverUrl) }
    }

    // ─── Внутренние методы ──────────────────────────────────────────────────────

    private fun resolveLogsDir(): File {
        val os = System.getProperty("os.name", "").lowercase()
        val base = when {
            os.contains("win") -> {
                val appData = System.getenv("APPDATA") ?: System.getProperty("user.home")
                File(appData, "Messenger")
            }
            else -> File(System.getProperty("user.home"), ".messenger")
        }
        return File(base, "logs")
    }

    private fun append(level: String, tag: String, message: String, stack: String?) {
        val file = logFile ?: return
        try {
            if (file.length() > MAX_SIZE) {
                file.renameTo(File(file.parentFile, "errors.log.old"))
                logFile = File(file.parentFile!!, "errors.log")
            }
            val ts = synchronized(fmt) { fmt.format(Date()) }
            val obj = buildJsonObject {
                put("timestamp", ts)
                put("level", level)
                put("tag", tag)
                put("message", message)
                put("platform", "desktop")
                put("os", System.getProperty("os.name", "unknown"))
                if (stack != null) put("stack", stack)
            }
            (logFile ?: file).appendText(json.encodeToString(obj) + "\n")
        } catch (e: Exception) {
            jul.severe("Failed to write log: ${e.message}")
        }
    }

    private fun installCrashHandler() {
        val prev = Thread.getDefaultUncaughtExceptionHandler()
        Thread.setDefaultUncaughtExceptionHandler { thread, throwable ->
            append(
                "error",
                "UncaughtException",
                "Crash in thread '${thread.name}': ${throwable.message}",
                throwable.stackTraceToString()
            )
            prev?.uncaughtException(thread, throwable)
        }
    }

    private fun doFlush(file: File, serverUrl: String) {
        try {
            val lines = file.readLines().filter { it.isNotBlank() }.take(FLUSH_LINES)
            if (lines.isEmpty()) return

            val entries = buildJsonArray {
                for (line in lines) {
                    runCatching {
                        val obj: JsonObject = json.parseToJsonElement(line).jsonObject
                        add(buildJsonObject {
                            put("timestamp", obj["timestamp"]?.jsonPrimitive?.content ?: "")
                            put("level",     obj["level"]?.jsonPrimitive?.content ?: "error")
                            put("userId",    "")
                            put("route",     "desktop/${obj["os"]?.jsonPrimitive?.content ?: ""}")
                            put("message",   "[${obj["tag"]?.jsonPrimitive?.content}] ${obj["message"]?.jsonPrimitive?.content}")
                            val stack = obj["stack"]?.jsonPrimitive?.content
                            if (!stack.isNullOrEmpty()) put("details", stack)
                        })
                    }
                }
            }
            if (entries.isEmpty()) return

            val body = json.encodeToString(buildJsonObject { put("entries", entries) })
                .toByteArray()

            val conn = URL("$serverUrl/api/client-errors").openConnection() as HttpURLConnection
            conn.requestMethod = "POST"
            conn.setRequestProperty("Content-Type", "application/json")
            conn.doOutput = true
            conn.outputStream.use { it.write(body) }

            if (conn.responseCode == HttpURLConnection.HTTP_NO_CONTENT) {
                val remaining = file.readLines().filter { it.isNotBlank() }.drop(lines.size)
                file.writeText(remaining.joinToString("\n") + if (remaining.isNotEmpty()) "\n" else "")
            }
            conn.disconnect()
        } catch (_: Exception) {
            // Сеть недоступна — логи сохранены, попытка при следующем старте.
        }
    }
}

File: apps/mobile/ios/Sources/Messenger/logger/AppErrorLogger.swift
// AppErrorLogger.swift — централизованное логирование ошибок iOS.
//
// Записывает структурированные JSON-строки в Documents/logs/errors.log.
// При превышении 5 МБ ротирует файл (→ errors.log.old).
// Перехватывает непойманные исключения через NSSetUncaughtExceptionHandler.
// При старте приложения отправляет накопленные логи на /api/client-errors.
import Foundation
import os

final class AppErrorLogger {

    static let shared = AppErrorLogger()

    private let maxSize: Int64 = 5 * 1024 * 1024 // 5 МБ
    private let flushLines = 100
    private let logger     = Logger(subsystem: "com.messenger", category: "errors")
    private let queue      = DispatchQueue(label: "com.messenger.errorlogger", qos: .background)
    private var logFile: URL?

    private init() {}

    // MARK: - Инициализация

    func initialize() {
        let logsDir = resolveLogsDir()
        try? FileManager.default.createDirectory(at: logsDir, withIntermediateDirectories: true)
        logFile = logsDir.appendingPathComponent("errors.log")
        installCrashHandler()
    }

    // MARK: - Публичный API

    func error(_ tag: String, _ message: String, _ error: Error? = nil) {
        logger.error("[\(tag)] \(message)\(error.map { ": \($0)" } ?? "")")
        append(level: "error", tag: tag, message: message,
               stack: error.map { "\($0)" })
    }

    func warn(_ tag: String, _ message: String) {
        logger.warning("[\(tag)] \(message)")
        append(level: "warn", tag: tag, message: message, stack: nil)
    }

    func info(_ tag: String, _ message: String) {
        logger.info("[\(tag)] \(message)")
        append(level: "info", tag: tag, message: message, stack: nil)
    }

    /// Отправить накопленные логи на сервер (вызвать при старте приложения).
    func flushToServer(serverUrl: String) {
        guard !serverUrl.isEmpty, let file = logFile else { return }
        queue.async { self.doFlush(file: file, serverUrl: serverUrl) }
    }

    // MARK: - Внутренние методы

    private func resolveLogsDir() -> URL {
#if canImport(UIKit)
        let docs = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first
            ?? URL(fileURLWithPath: NSHomeDirectory())
        return docs.appendingPathComponent("logs")
#else
        // macOS (не используется при swift test, но нужен для компиляции)
        return URL(fileURLWithPath: NSHomeDirectory())
            .appendingPathComponent(".messenger/logs")
#endif
    }

    private func append(level: String, tag: String, message: String, stack: String?) {
        queue.async { [weak self] in
            guard let self, let fileUrl = self.logFile else { return }
            do {
                // Ротация при превышении лимита.
                if let attrs = try? FileManager.default.attributesOfItem(atPath: fileUrl.path),
                   let size = attrs[.size] as? Int64, size > self.maxSize {
                    let old = fileUrl.deletingLastPathComponent()
                        .appendingPathComponent("errors.log.old")
                    try? FileManager.default.removeItem(at: old)
                    try? FileManager.default.moveItem(at: fileUrl, to: old)
                }

                let ts = ISO8601DateFormatter().string(from: Date())
                var obj: [String: Any] = [
                    "timestamp": ts,
                    "level":     level,
                    "tag":       tag,
                    "message":   message,
                    "platform":  "ios"
                ]
                if let stack { obj["stack"] = stack }

                let data = try JSONSerialization.data(withJSONObject: obj)
                var line = data
                line.append(contentsOf: [0x0A]) // "\n"

                if FileManager.default.fileExists(atPath: fileUrl.path) {
                    let handle = try FileHandle(forWritingTo: fileUrl)
                    handle.seekToEndOfFile()
                    handle.write(line)
                    handle.closeFile()
                } else {
                    try line.write(to: fileUrl)
                }
            } catch {
                self.logger.error("Failed to write log: \(error)")
            }
        }
    }

    private func installCrashHandler() {
        NSSetUncaughtExceptionHandler { exception in
            AppErrorLogger.shared.append(
                level: "error",
                tag:   "UncaughtException",
                message: exception.name.rawValue + ": " + (exception.reason ?? ""),
                stack:   exception.callStackSymbols.joined(separator: "\n")
            )
            // Небольшая пауза, чтобы запись успела завершиться.
            Thread.sleep(forTimeInterval: 0.3)
        }
    }

    private func doFlush(file: URL, serverUrl: String) {
        guard let content = try? String(contentsOf: file, encoding: .utf8) else { return }
        let lines = content.components(separatedBy: "\n")
            .filter { !$0.isEmpty }
            .prefix(flushLines)
        guard !lines.isEmpty else { return }

        var entries: [[String: Any]] = []
        for line in lines {
            guard let data = line.data(using: .utf8),
                  let obj  = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
            else { continue }
            var entry: [String: Any] = [
                "timestamp": obj["timestamp"] ?? "",
                "level":     obj["level"]     ?? "error",
                "userId":    "",
                "route":     "ios",
                "message":   "[\(obj["tag"] ?? "")] \(obj["message"] ?? "")"
            ]
            if let stack = obj["stack"] as? String, !stack.isEmpty {
                entry["details"] = stack
            }
            entries.append(entry)
        }
        guard !entries.isEmpty else { return }

        guard let body = try? JSONSerialization.data(withJSONObject: ["entries": entries]),
              let url  = URL(string: "\(serverUrl)/api/client-errors")
        else { return }

        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = body

        let sem = DispatchSemaphore(value: 0)
        URLSession.shared.dataTask(with: req) { _, resp, _ in
            defer { sem.signal() }
            if (resp as? HTTPURLResponse)?.statusCode == 204 {
                // Удалить отправленные строки.
                let allLines = (try? String(contentsOf: file, encoding: .utf8))?
                    .components(separatedBy: "\n").filter { !$0.isEmpty } ?? []
                let remaining = Array(allLines.dropFirst(lines.count))
                let newContent = remaining.isEmpty ? "" : remaining.joined(separator: "\n") + "\n"
                try? newContent.write(to: file, atomically: true, encoding: .utf8)
            }
        }.resume()
        sem.wait()
    }
}
